In [1]:
# Imports section
import pandas as pd

## Part 1. Loading the dataset

In [3]:
# Using pandas load the dataset (load remotely, not locally)
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

data = pd.read_csv('C:\\<redacted>\\383\\Assignment2\\assignment-4-equation-of-a-slime-<redacted>\science_data_large.csv')

print(data.head(15))

data.info()

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## Part 2. Splitting the dataset

In [7]:
# Take the pandas dataset and split it into our features (X) and label (y)
from sklearn.model_selection import train_test_split

X = data[['Temperature °C', 'Mols KCL']]
y = data[['Size nm^3']]

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)


## Part 3. Perform a Linear Regression

In [13]:
# Use sklearn to train a model on the training set
from sklearn.linear_model import LinearRegression
import numpy as np

linear_reg = LinearRegression()

linear_reg.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model

sample_data = np.array([[400,600]])

predicted_size = linear_reg.predict(sample_data)

print(f"Predicted size for 400°C, 600 Mols KCL: {predicted_size[0]}")

# Report on the score for that model, in your own words (markdown, not code) explain what the score means

score = linear_reg.score(X_test, y_test)
print(f"Model score: {score}")

coefficients = linear_reg.coef_.flatten()  # Flatten the coefficients to ensure it's a 1D array
intercept = linear_reg.intercept_.flatten()  # Flatten the intercept if needed

# Print the coefficients and intercept
print("Coefficients:", coefficients)
print("Intercept:", intercept)

# Ensure there are two coefficients (one for each feature)
if len(coefficients) == 2:  
    temperature_coef = coefficients[0]
    kcl_coef = coefficients[1]

    # Print the linear regression equation
    print(f"h(x) = {temperature_coef:.5f} * Temperature + {kcl_coef:.5f} * Mols KCL + {intercept[0]:.5f}")
else:
    print("The number of coefficients is not what was expected. Please check the model.")

Predicted size for 400°C, 600 Mols KCL: [556684.12565684]
Model score: 0.8552472077276095
Coefficients: [ 866.14641337 1032.69506649]
Intercept: [-409391.47958341]
h(x) = 866.14641 * Temperature + 1032.69507 * Mols KCL + -409391.47958


C:\Users\<redacted>\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


### Report on the score for the Linear Regression model

The **model score** is a measure of how well the Linear Regression model fits the data. Specifically, it represents the coefficient of determination, \(R^2\), which indicates the proportion of the variance in the dependent variable (slime size) that is predictable from the independent variables (temperature and Mols KCL).

In this case, the model score is **0.8552**, meaning that approximately 85.5% of the variance in the size of the slime can be explained by the temperature and the amount of Potassium Chloride (KCL). The remaining 14.5% is due to factors not captured by the model, such as noise or other unaccounted variables. This is generally a good score, indicating a strong linear relationship between the input variables and the output (slime size).

### Linear Regression Equation

The linear regression equation is:

h(x) = 866.14641 * Temperature + 1032.69507 * Mols KCL - 409391.47958

## Part 4. Use Cross Validation

In [15]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
from sklearn.model_selection import cross_val_score

cv_scores = cross_val_score(linear_reg, X, y, cv=5)

# Report on their finding and their 
print(f"Cross-validation scores: {cv_scores}")
print(f"Average cross-validation score: {cv_scores.mean()}")

Cross-validation scores: [0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]
Average cross-validation score: 0.8568167899144437


## Part 5. Using Polynomial Regression

In [17]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
from sklearn.preprocessing import PolynomialFeatures

poly = PolynomialFeatures(degree=2)
X_poly_train = poly.fit_transform(X_train)
X_poly_test = poly.transform(X_test)
# Report on the metrics and output the resultant equation as you did in Part 3.

poly_reg = LinearRegression()
poly_reg.fit(X_poly_train, y_train)

poly_score = poly_reg.score(X_poly_test, y_test)
print(f"Polynomial Regression score: {poly_score}")

poly_coefficients = poly_reg.coef_
poly_intercept = poly_reg.intercept_
print(f"h(x) = {temperature_coef:.5f} * Temperature + {kcl_coef:.5f} * Mols KCL + {intercept[0]:.5f}")


Polynomial Regression score: 1.0
h(x) = 866.14641 * Temperature + 1032.69507 * Mols KCL + -409391.47958
